---
title: "Chemo.05.GenomicTraits"
author: "Angel Rain & Sara Beier"
date: "9/9/2022"
output:
  pdf_document:
    number_sections: yes
    toc: yes
  html_document:
    toc: yes
    df_print: paged
---
```{r, include=F,message=F}
library(knitr)
opts_chunk$set(tidy.opts=list(width.cutoff=80),tidy=TRUE)
```

## Load packages

```{r setup, include=T,message=F,warning=F}
rm(list=ls())
library(phyloseq)
library(reshape2)
library(dplyr)
library(ggplot2)
library(vegan)
library(rstatix) #Homogenety of variance test
library(olsrr) # test_normality
library(egg) # Tag figures
library(rstatix) #Tibble function
library(microViz) #TO use re_order function
library(kableExtra) #Table format

#SARA: please check, keep only libraries that are needed in the script
#######################################
cbbPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
```

# Setting up workspace
## Load metadata
```{r}
# Create metadata for experimental setup
schema<-data.frame(sample.ID=paste0("C10-",rep(1:9,each=12),"-",sprintf("%02d",1:12)),
                   T=rep(1:9,each=12),
                   Chem.ID=sprintf("%02d",1:12),
                   DOM=rep(c("L","H"),each=6),
                   Sal=c("C","D"))

tibble(schema)
```

#Loading phyloseq object with ASV count table from #dada2
```{r}
#Loading phyloseq object from #dada2
ps <- readRDS("../data/dada2.output/chem.ps.rds")
#Phyloseq object contain abundance table, sample information, taxanomic information and the phylogenetic tree

# Loadgin phylogenetic tree
chem.tree=read_tree("../data/dada2.output/dada-chem.GTR2")
phy_tree(ps)<-chem.tree #Adding phylo-tree to the phyloseq object

#Phyloseq object contain abundance table, sample information, taxonomic information and the phylogenetic tree
ps 
```

## Load predicted traits
```{r}
#Resilience related genes
#load RRN predicted from rrnDB tree and trait data
pic.16s.custom <- read.table("../data/picrust2/trait.predicted/pic.chemo10.16S_predicted_custom_tree.txt", header=T) 
tibble(pic.16s.custom)
#load generation time predicted from PICRUST2 default tree and database
pic.d.gRodon.default <- read.table("../data/picrust2/trait.predicted/pic.d.gRodon.retransformed.txt", header=T)  
tibble(pic.d.gRodon.default)

#Resistance-related genes
#load %TF predicted from PICRUSt2 default tree and database
pic.TFr.default <- read.table("../data/picrust2/trait.predicted/pic.TF_perc.retransformed.txt", header=T)
tibble(pic.TFr.default)
#load genome size predicted from PICRUSt2 default tree and database
pic.gs.default <- read.table("../data/picrust2/trait.predicted/pic.genome.size.retransformed.txt", header=T)
tibble(pic.gs.default)

```

# Community indexes
## Calculation of the Community weigthed mean (CWM)
CWMs were obtained by summing predicted and abundance-weighted trait-values for all ASVs in each community

### Relative abundance data
```{r}
#Rarefy by minimum read numbers and transform to relative data
ps=rarefy_even_depth(ps, min(rowSums(otu_table(ps))),rngseed=1,replace=F,trimOTUs=F)

#Estimating relative abundance
rOTUdf.rar <- prop.table(otu_table(ps),1)

# New phyloseq-project with rarefied ASV table
otu_table(ps) <- otu_table(rOTUdf.rar, taxa_are_rows=FALSE) 
ps 

# Keep ASVs with prevalence equivalent to more 0 reads
ps<-prune_taxa(taxa_sums(ps)>0,ps) 
ps
# Setting up metadata

# Samples in phyloseq object did not correspond to the metadata (schema), so we proceed to reorder ps-data base in the schema$sample.ID ##SARA: ???; samples from chem3?
new_order <- schema$sample.ID
ps=ps %>%
  ps_reorder(new_order) #MicroViz package

# Extract ASV count table
counts=t(otu_table(ps))

```


### Remove ASVs without close relatives in the default reference database (NSTI<1)
```{r}
counts.s.default <- counts[row.names(counts) %in% pic.gs.default[pic.gs.default$metadata_NSTI<1,1], ] #extract ASVs with NSTI<1 in default reference database
colSums (counts.s.default) #check which proportion of sequences is left after removing ASVs with NSTI<1
min(colSums (counts.s.default))
counts.s.rel.default <- as.data.frame.matrix(prop.table(t(t(counts.s.default)),2)) #re-normalize remaining count-data
colSums (counts.s.rel.default) #should sum up again to 1
```
### Remove ASVs without close relatives in the custom reference database (NSTI<1)
```{r}

counts.s.custom <- counts[row.names(counts) %in% pic.16s.custom[pic.16s.custom$metadata_NSTI<1,1], ] #extract ASVs with NSTI<1 (= ASVs with no close relative in the picrust2 reference database)
colSums (counts.s.custom) #check which proportion of sequences is left after removing ASVs with NSTI<1
min(colSums (counts.s.custom))
counts.s.rel.custom <- as.data.frame.matrix(prop.table(t(t(counts.s.custom)),2)) #re-normalize remaining count-data
colSums (counts.s.rel.custom) #should sum up again to 1


```
## Estimate alpha diversity (Shannon diversity index)
```{r}
#Shannon diversity
H<- diversity(counts, index="shannon", MARGIN=2, base = exp(1)) 
tibble(H)
```

## Community weigthed means (CWMs)
For each sample and genomic trait (16S rRNA  gene copy number, generation time, %transcription factors ,and generation time), the community weighted mean (CWM) was used for downstream statistical analyses.

```{r}
## 16s rRNA gene copy number
counts.16s <- merge(pic.16s.custom,counts.s.rel.custom, by.x='sequence', by.y=0)
row.names(counts.16s) <- counts.16s[,1]
counts.16s <-counts.16s[,c(2,4:dim(counts.16s)[2])]
# CWM 16S rRNA gene copy per sample
av.16s <- colSums(counts.16s[,1]*counts.16s[,2:dim(counts.16s)[2]])
av.16s[1:10]

## Generation time gRodon (from codon usage bias using the gRodon R package)
counts.generationstime.gR <- merge(pic.d.gRodon.default,counts.s.rel.default, by.x='sequence', by.y=0)
row.names(counts.generationstime.gR) <- counts.generationstime.gR[,1]
# Select sample columns
counts.generationstime.gR <-counts.generationstime.gR[,c(2,4:dim(counts.generationstime.gR)[2])]
# CWM generation time
av.dgR <- colSums(counts.generationstime.gR[,1]*counts.generationstime.gR[,2:dim(counts.generationstime.gR)[2]]) 
av.dgR[1:10]

# Percent transcription factors (%TF)
counts.TFr <- merge(pic.TFr.default,counts.s.rel.default, by.x='sequence', by.y=0)#create a column with predicted genome size for each ASV
row.names(counts.TFr) <- counts.TFr[,1]
# Select sample columns
counts.TFr <-counts.TFr[,c(2,4:dim(counts.TFr)[2])]
# CWM generation time
av.TFr <- colSums(counts.TFr [,1]*counts.TFr [,2:dim(counts.TFr)[2]])
av.TFr[1:10]

## Genome size (in Mbp)
counts.gs <- merge(pic.gs.default,counts.s.rel.default, by.x='sequence', by.y=0)
row.names(counts.gs) <- counts.gs[,1]
#Select sample columns
counts.gs <-counts.gs[,c(2,4:dim(counts.gs)[2])]
# CWM Genome size
av.gs <- colSums(counts.gs[,1]*counts.gs[,2:dim(counts.gs)[2]]) 
av.gs[1:10]


#NSTI custom
counts.NSTIs <- merge(pic.16s.custom,counts.s.rel.custom, by.x='sequence', by.y=0) #create a column with predicted 16s copy numbers for each ASV
row.names(counts.NSTIs) <- counts.NSTIs[,1]
counts.NSTIs <-counts.NSTIs[,c(3,4:dim(counts.NSTIs)[2])] #select releavnt samples
av.NSTI <- colSums(counts.NSTIs[,1]*counts.NSTIs[,2:dim(counts.NSTIs)[2]]) #average number 0f 16s rRNA gene copy per sample
summary(av.NSTI)

#NSTI default
counts.NSTIs <- merge(pic.gs.default,counts.s.rel.default, by.x='sequence', by.y=0) #create a column with predicted 16s copy numbers for each ASV
row.names(counts.NSTIs) <- counts.NSTIs[,1]
counts.NSTIs <-counts.NSTIs[,c(3,4:dim(counts.NSTIs)[2])] #select releavnt samples
av.NSTI <- colSums(counts.NSTIs[,1]*counts.NSTIs[,2:dim(counts.NSTIs)[2]]) #average number 0f 
summary(av.NSTI)
```

# Community trait distribution during the experiment
## Dataframe and format trait CWM values
```{r}
# Data frame with CWM trait data and sample schema
traits<-cbind(schema,av.16s,av.gs,av.dgR,av.TFr,H) 

# Formatting data set from wide to long format
traits.w <-melt(traits[,2:10], id.vars = c("Sal", "DOM", 'T'),
                measure.vars = c("av.16s", "av.dgR","av.TFr","av.gs","H"))

#Add column with Replicate ID
traits.w$Rep=rep(c("1","2","3"),each=2)
traits.w$Rep=as.factor(traits.w$Rep)

#Add Column with sample time (day)
traits.w$Time=as.numeric(rep(c(4,8,15,18,22,29,36,39,41),each=12))
```

## Summaring data replicate mean values
```{r}
tibble(aggregate(value~Sal+DOM+variable,traits.w,mean))
```

## Test for normality and homogeneity of variances
```{r}
#Normality Kolmogovor smirnov test
l=length(levels(traits.w$variable))
traits.w$T=factor(traits.w$T)
sum.normality=data.frame(variable=rep(NA,l),L_C=rep(NA,l),L_D=rep(NA,l),
                         H_C=rep(NA,l),H_D=rep(NA,l))

for (i in 1:length(levels(traits.w$variable))){
tmp=traits.w[traits.w$variable==levels(traits.w$variable)[i],]
sum.normality$variable[i]=levels(traits.w$variable)[i]
sum.normality$L_C[i]=ols_test_normality((tmp$value[tmp$DOM=="L" & tmp$Sal=="C"]))[[1]][[2]]
sum.normality$L_D[i]=ols_test_normality((tmp$value[tmp$DOM=="L" & tmp$Sal=="D"]))[[1]][[2]]
sum.normality$H_C[i]=ols_test_normality((tmp$value[tmp$DOM=="H" & tmp$Sal=="C"]))[[1]][[2]]
sum.normality$H_D[i]=ols_test_normality((tmp$value[tmp$DOM=="H" & tmp$Sal=="D"]))[[1]][[2]]
}
sum.normality[,2:5]=round(sum.normality[,2:5],3)
tibble(sum.normality)

# Homogeneity of variances
HV=traits.w %>%
  group_by(variable,DOM,Sal) %>%
  levene_test(value~T)
tibble(HV)
```

## Repeated measurement ANOVA
A repeated measurement anova was applied seperately for the two DOM regimes to test the effect of the disturbance regime on the distribution of the resilience- and resistance-related genomic traits.
```{r}
#Repeated measurements ANOVA for LDOM

list.rm_anova=list()
m.rm_anova=data.frame(variable=rep(NA,l),F_Time=rep(NA,l),P_Time=rep(NA,l),
                      F_Sal=rep(NA,l),P_Sal=rep(NA,l))
for (i in 1:length(levels(traits.w$variable))){
  list.rm_anova[[i]] <- with(traits.w[traits.w$DOM=="L" & traits.w$variable==levels(traits.w$variable)[i],],
                             aov(value ~ T*Sal +Error(Rep)))
  m.rm_anova$variable[i]=levels(traits.w$variable)[i]
  m.rm_anova$F_Time[i]=unlist(summary(list.rm_anova[[i]]))['Error: Within.F value1']
  m.rm_anova$P_Time[i]=unlist(summary(list.rm_anova[[i]]))['Error: Within.Pr(>F)1']
  m.rm_anova$F_Sal[i]=unlist(summary(list.rm_anova[[i]]))['Error: Within.F value2']
  m.rm_anova$P_Sal[i]=unlist(summary(list.rm_anova[[i]]))['Error: Within.Pr(>F)2']
}

m.rm_anova[,2:5]=round(m.rm_anova[,2:5],3)
a.rm_anova <- m.rm_anova
tibble(a.rm_anova)

#Repeated measurements ANOVA for HDOM
list.rm_anova=list()
m.rm_anova=data.frame(variable=rep(NA,l),F_Time=rep(NA,l),P_Time=rep(NA,l),
                      F_Sal=rep(NA,l),P_Sal=rep(NA,l))
for (i in 1:length(levels(traits.w$variable))){
  list.rm_anova[[i]] <- with(traits.w[traits.w$DOM=="H" & traits.w$variable==levels(traits.w$variable)[i],],
                             aov(value ~ T*Sal+Error(Rep)))
  m.rm_anova$variable[i]=levels(traits.w$variable)[i]
  m.rm_anova$F_Time[i]=unlist(summary(list.rm_anova[[i]]))['Error: Within.F value1']
  m.rm_anova$P_Time[i]=unlist(summary(list.rm_anova[[i]]))['Error: Within.Pr(>F)1']
  m.rm_anova$F_Sal[i]=unlist(summary(list.rm_anova[[i]]))['Error: Within.F value2']
  m.rm_anova$P_Sal[i]=unlist(summary(list.rm_anova[[i]]))['Error: Within.Pr(>F)2']
}

m.rm_anova[,2:5]=round(m.rm_anova[,2:5],3)
a.rm_anova <- m.rm_anova
tibble(a.rm_anova)
```
# Paired-test per Genomic trait
```{r}

traits.w.mean=aggregate(value~Sal+DOM+T+variable,data=traits.w,mean)
res.ttest=list()
res.ttest.df=data.frame(variable=levels(traits.w.mean$variable),direction=c("greater","less","greater","greater","greater"),
                        LDOM.pvalue=NA,HDOM.pvalue=NA)

for (i in 1:length(levels(traits.w.mean$variable))){
  tmp=traits.w.mean[traits.w.mean$variable==levels(traits.w$variable)[i],]
  value.control=tmp[tmp$DOM=="L" & tmp$Sal=="C",]
  value.disturbance=tmp[tmp$DOM=="L" & tmp$Sal=="D",]
res.ttest[[i]]=t.test(value.disturbance$value,value.control$value,alternative =res.ttest.df$direction[i] ,var.equal=T,paired = T)
res.ttest.df$LDOM.pvalue[i]=res.ttest[[i]]$p.value
}


res.ttest=list()
for (i in 1:length(levels(traits.w.mean$variable))){
  tmp=traits.w.mean[traits.w.mean$variable==levels(traits.w$variable)[i],]
  value.control=tmp[tmp$DOM=="H" & tmp$Sal=="C",]
  value.disturbance=tmp[tmp$DOM=="H" & tmp$Sal=="D",]
res.ttest[[i]]=t.test(value.disturbance$value,value.control$value,alternative = res.ttest.df$direction[i],var.equal=T,paired = T)
res.ttest.df$HDOM.pvalue[i]=res.ttest[[i]]$p.value
}

tibble(res.ttest.df)
```
## Manuscript Figure 3
```{r, fig.dim=c(5.5,6)}
# New facet label names for dose variable
bxp_labs <-c("" ,"",""," ","")
names(bxp_labs) <- levels(traits.w$variable)
traits.w$DOM=factor(traits.w$DOM,levels=c("L","H"))

levels(traits.w$T)=c("4","8","15","18","22","29","36","39","41")

bxp=traits.w%>%
  ggplot(aes(x =T, y =value,colour=Sal)) +
  geom_boxplot(aes(colour=(Sal)),outlier.shape =NA,alpha=0.3,size=0.4)+
  geom_jitter(aes(colour = Sal),shape=21,size=.5,position = position_jitterdodge())+
  scale_colour_manual(values=cbbPalette,name="")+
  theme_bw()+ylab("")+ scale_y_continuous(expand = expansion(mult = c(0, .25)))+
  theme(panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
        axis.text.x = element_text(size=10), axis.text.y = element_text(size=10))+
  theme(legend.position=c(0.9,1.02),legend.direction="horizontal",
        legend.key=element_blank(),legend.background=element_blank())+
  theme(text=element_text(size=10, family="ArialMT"))+
  facet_grid(variable~DOM,scale="free_y",switch = "y", labeller = labeller(variable =              bxp_labs))+
  xlab("Time (days)")+ theme(strip.placement.y = "outside",
        strip.text.y = element_text(angle = 270),
        strip.background = element_blank())+
  labs(tag = "LDOM                                                HDOM") +
  theme(plot.tag.position = c(0.38, 1.02))

#Labels using facet_tag
bxp=tag_facet(bxp,open = "", close = "",
              tag_pool=c(" A) RNN", " B) RNN "," C) d (hrs)"," D) d (hrs)"," E) TF (%)"," F) TF (%)"," G) Genome Size (Mbp)", " H) Genome Size (Mbp)"," I) Shannon"," J) Shannon"),x=0, fontface = 1,size=3,hjust=0)
bxp=bxp+theme(plot.margin=margin(t = 20, r = 5, b = 5, l = 5, unit = "pt"))
bxp

```

Boxplots displaying CWMs of genomic traits. LDOM and HDOM in the left and right panels restively for A, B), RRN,  C, D) generation time (d), E, F), %TF G, H) genome size and I, J) Shannon diversity index.

```{r, include=F}
pdf("../figures/Figure3_genomic.traits.pdf",height=7, width=6)
bxp
dev.off()
```




